### DISCUSSION TITLE
Program crashing when reading pdf with readImages

### DISCUSSION THREAD

**Discussion by tsftd** [ImageMagick/ImageMagick#6928] [Help] [answered]
### ImageMagick version

ImageMagick-7.1.1-Q16-HDRI

### Operating system, version and so on

Win10, MSVS 2022

### Description

EDIT:
So I finally figured it out: I needed ghostscript.
I had (again, many years ago) added a ghostscript install and hardcoded the relative location in the delegates file in order to make it portable to machines without gs installed.
will leave this up in case it helps someone else.
/EDIT

So I have a simple little program that reads in a PDF and converts it to png/jpg, based on a few settings (DPI, etc). Nothing fancy, just a tool that I use for work. Tried using it today and it kept crashing without error. No problem, it's been a good 5+ years since I last updated it and I've since not only moved to a new ImageMagick version, but made some other changes like removing Boost from all of my projects since the new C++ standard supports all of its features that I need. (turns out that the old version is actually fine, there was just a kanji in the filepath which borked it since I don't have proper support for it but that's another subject)

Did the usual stuff updating the project (converting all boost functions to standard c++, etc), finally got it to compile after a few tweaks.

Crashes on Magick::readimages() call. That section of code is unchanged from the original, so I have no clue why it would pose a problem now when the old version works just fine. Code follows, my apologies for the atrocious style:

new version:
<pre>void ReadPDF(string& pdf,ReadOptions& options,string& outpath){
	if(silent==false){
		lockdisp.lock();
		cout << "Reading \""+pdf+"\"" << endl;
		lockdisp.unlock();
	}
	std::list<Image> images;
	std::list<Image>::iterator position;
	shared_mutex lockwrite;
	Magick::readImages(&images,pdf,options);
	pdf="\""+pdf+"\"";
	if(silent==false){
		lockdisp.lock();
		cout << "Finished reading "+pdf << endl;
		lockdisp.unlock();
	}
	position=images.begin();
	std::filesystem::path temppath=pdf;
	outpath=outpath+temppath.stem().string()+" - image ";
	string footer="/"+to_string(images.size())+": Written";
	pdf=pdf+" Page ";
	for(int_fast16_t page=1;position!=images.end();page++,position++){
		lockwrite.lock_shared();
		thread WritePaget(WritePage,std::ref(*position),outpath+to_string(page)+"."+ext,pdf+to_string(page)+footer,std::ref(lockwrite));
		WritePaget.detach();
	}
	lockwrite.lock();
	lockwrite.unlock();
	lockpdf.unlock_shared();
}</pre>
old version, for reference (identical, except for the changes to the boost functions, which are all after the error):
<pre>void ReadPDF(string& pdf,ReadOptions& options,string& outpath){
	if(silent==false){
		lockdisp.lock();
		cout << "Reading \""+pdf+"\"" << endl;
		lockdisp.unlock();
	}
	std::list<Image> images;
	std::list<Image>::iterator position;
	shared_mutex lockwrite;
	readImages(&images,pdf,options);
	pdf="\""+pdf+"\"";
	if(silent==false){
		lockdisp.lock();
		cout << "Finished reading "+pdf << endl;
		lockdisp.unlock();
	}
	position=images.begin();
	std::filesystem::path temppath=pdf;
	outpath=outpath+temppath.stem().string()+" - image ";
	string footer="/"+lexical_cast<string>(images.size())+": Written";
	pdf=pdf+" Page ";
	for(int_fast16_t page=1;position!=images.end();page++,position++){
		lockwrite.lock_shared();
		boost::thread WritePage(WritePage,boost::ref(*position),outpath+lexical_cast<string>(page)+"."+ext,pdf+lexical_cast<string>(page)+footer,boost::ref(lockwrite));
	}
	lockwrite.lock();
	lockwrite.unlock();
	lockpdf.unlock_shared();
}</pre>
*note, the function is threaded, however I have removed the threading to eliminate that as an issue, and the problem persists.

the only option that I change from default is: options.density(Geometry(150,150));

program output is:
<pre>Density option is missing, and has been set to the default of 150 ppi
Extension option is missing, and has been set to the default of "png"
Trim option is missing, and has been set to the default of "false"
Fuzz option is missing, and has been left at the Magick++ default
Crop option is missing, and has been set to the default of "0,0,0,0" (no cropping)
Reading "D:\NextCloud\ske\cold.pdf"</pre>

Call stack is:
<pre> 	ucrtbase.dll!00007ffad821f61e()	Unknown
 	ucrtbase.dll!00007ffad821ed39()	Unknown
 	vcruntime140_1.dll!00007ffab9bf1ab1()	Unknown
 	vcruntime140_1.dll!00007ffab9bf232f()	Unknown
 	vcruntime140_1.dll!00007ffab9bf2389()	Unknown
 	vcruntime140_1.dll!00007ffab9bf4189()	Unknown
 	ntdll.dll!00007ffadac93cff()	Unknown
 	ntdll.dll!00007ffadac0e456()	Unknown
 	ntdll.dll!00007ffadac92cee()	Unknown
 	KernelBase.dll!00007ffad83a536c()	Unknown
 	vcruntime140.dll!00007ffab65e6ba7()	Unknown
 	CORE_RL_Magick++_.dll!00007ffaafc742d8()	Unknown
>	[Inline Frame] 2023-unPDFit.exe!Magick::readImages(std::list<Magick::Image,std::allocator<Magick::Image>> *) Line 2750	C++
 	[Inline Frame] 2023-unPDFit.exe!Magick::readImages(std::list<Magick::Image,std::allocator<Magick::Image>> *) Line 2757	C++
 	2023-unPDFit.exe!ReadPDF(std::string & pdf, Magick::ReadOptions & options, std::string & outpath) Line 57	C++
 	[External Code]	</pre>

in case it's not obvious, i'm *not* a professional coder, but I noticed the [Inline Frame] above and disabled inlining, and got slightly more information:
<pre> 	ucrtbase.dll!00007ffad821f61e()	Unknown
 	ucrtbase.dll!00007ffad821ed39()	Unknown
 	vcruntime140_1.dll!00007ffab9bf1ab1()	Unknown
 	vcruntime140_1.dll!00007ffab9bf232f()	Unknown
 	vcruntime140_1.dll!00007ffab9bf2389()	Unknown
 	vcruntime140_1.dll!00007ffab9bf4189()	Unknown
>	2023-unPDFit.exe!__GSHandlerCheck_EH4(_EXCEPTION_RECORD * ExceptionRecord, void * EstablisherFrame, _CONTEXT * ContextRecord, _DISPATCHER_CONTEXT * DispatcherContext) Line 86	C++
 	ntdll.dll!00007ffadac93cff()	Unknown
 	ntdll.dll!00007ffadac0e456()	Unknown
 	ntdll.dll!00007ffadac92cee()	Unknown
 	KernelBase.dll!00007ffad83a536c()	Unknown
 	vcruntime140.dll!00007ffab65e6ba7()	Unknown
 	CORE_RL_Magick++_.dll!00007ffaafc742d8()	Unknown
 	2023-unPDFit.exe!Magick::readImages<std::list<Magick::Image,std::allocator<Magick::Image>>>(std::list<Magick::Image,std::allocator<Magick::Image>> * sequence_, const std::string & imageSpec_, Magick::ReadOptions & options) Line 2750	C++
 	2023-unPDFit.exe!ReadPDF(std::string & pdf, Magick::ReadOptions & options, std::string & outpath) Line 58	C++
 	[External Code]</pre>

it's not exactly critical, as like I said the original version is actually working, but if I ever want to update it, I need to figure out what the problem is, so any help would be appreciated.

**Comment by snibgo** [comment, marked as answer]
As you probably know, IM delegates the rasterization of PDFs to Ghostscript. Perhaps that is where the problem is occurring.

You say:
> Tried using it today and it kept crashing without error.

You don't seem to have any error checking or reporting. If `readImages()` fails, your program would seem to continue regardless.

**Reply by tsftd** [reply]
indeed, i didn't have a try/catch block on it. that's what led me to figuring out that gs was the issue.

it does crash without the try/catch block though (code posted above will crash, not continue)

**Reply by tsftd** [reply]
also, thanks for the community support! i poke my head in here (or rather the old forums) every few years when something stumps me, and always recognize your name helping me
